File: apps/api/src/dashboards/dashboards.service.ts
import { ForbiddenException, Injectable } from "@nestjs/common";
import { PrismaService } from "../common/prisma/prisma.service";
import { AuthenticatedUser } from "../common/decorators/current-user.decorator";

@Injectable()
export class DashboardsService {
  constructor(private readonly prisma: PrismaService) {}

  /** Part K tenant dashboard: bookings, payments, agreements, support tickets — theirs only. */
  async tenantDashboard(userId: string) {
    const [bookings, supportTickets, favouritesCount, savedSearchesCount] = await Promise.all([
      this.prisma.booking.findMany({
        where: { userId },
        orderBy: { createdAt: "desc" },
        include: {
          unit: { include: { property: true } },
          payments: true,
          agreements: { include: { versions: { orderBy: { version: "desc" }, take: 1 } } },
          installments: true,
          // So the "leave a review" prompt only shows for a COMPLETED booking
          // that doesn't already have one (see MyBookingsPage).
          reviews: { select: { id: true } },
        },
      }),
      this.prisma.supportTicket.findMany({
        where: { userId },
        orderBy: { createdAt: "desc" },
        take: 5,
      }),
      this.prisma.favourite.count({ where: { userId } }),
      this.prisma.savedSearch.count({ where: { userId } }),
    ]);

    return {
      bookings,
      supportTickets,
      counts: {
        activeBookings: bookings.filter((b: { status: string }) => ["PENDING_PAYMENT", "CONFIRMED"].includes(b.status))
          .length,
        favourites: favouritesCount,
        savedSearches: savedSearchesCount,
      },
    };
  }

  /**
   * Part K manager dashboard: portfolio composition, verification pipeline
   * status, recent bookings, and revenue — scoped to one organisation.
   * Reuses the same membership check as PropertiesService so a manager can
   * never see another organisation's numbers.
   */
  async managerDashboard(user: AuthenticatedUser, organisationId: string) {
    if (!user.roles.includes("Admin")) {
      const membership = await this.prisma.organisationMember.findUnique({
        where: { organisationId_userId: { organisationId, userId: user.userId } },
      });
      if (!membership) {
        throw new ForbiddenException("You do not have access to this organisation's dashboard.");
      }
    }

    const [propertiesByStatus, properties, recentBookings] = await Promise.all([
      this.prisma.property.groupBy({
        by: ["publicationStatus"],
        where: { organisationId },
        _count: { _all: true },
      }),
      this.prisma.property.findMany({
        where: { organisationId },
        select: { id: true, units: { select: { id: true } } },
      }),
      this.prisma.booking.findMany({
        where: { unit: { property: { organisationId } } },
        orderBy: { createdAt: "desc" },
        take: 10,
        include: { unit: { include: { property: true } }, payments: true },
      }),
    ]);

    const totalUnits = properties.reduce((sum: number, p: { units: unknown[] }) => sum + p.units.length, 0);

    const revenueResult = await this.prisma.payment.aggregate({
      where: { status: "SUCCEEDED", booking: { unit: { property: { organisationId } } } },
      _sum: { amount: true },
    });

    return {
      propertiesByStatus: Object.fromEntries(
        propertiesByStatus.map((row: { publicationStatus: string; _count: { _all: number } }) => [row.publicationStatus, row._count._all])
      ),
      totalProperties: properties.length,
      totalUnits,
      recentBookings,
      totalRevenue: revenueResult._sum.amount ?? 0,
    };
  }

  /** Part K admin dashboard: platform-wide operational view. */
  async adminDashboard() {
    const [
      propertiesByStatus,
      flaggedConflicts,
      verificationQueueCount,
      bookingsByStatus,
      totalUsers,
      totalRevenue,
      refundsPendingDualControl,
      supportTicketsByPriority,
    ] = await Promise.all([
      this.prisma.property.groupBy({ by: ["publicationStatus"], _count: { _all: true } }),
      this.prisma.property.count({ where: { conflictStatus: "FLAGGED" } }),
      this.prisma.property.count({ where: { publicationStatus: "REVIEW" } }),
      this.prisma.booking.groupBy({ by: ["status"], _count: { _all: true } }),
      this.prisma.user.count(),
      this.prisma.payment.aggregate({ where: { status: "SUCCEEDED" }, _sum: { amount: true } }),
      this.prisma.refund.count({ where: { requiresDualControl: true, approvedBy: null } }),
      this.prisma.supportTicket.groupBy({
        by: ["priority"],
        where: { status: { in: ["OPEN", "IN_PROGRESS"] } },
        _count: { _all: true },
      }),
    ]);

    return {
      propertiesByStatus: Object.fromEntries(
        propertiesByStatus.map((row: { publicationStatus: string; _count: { _all: number } }) => [row.publicationStatus, row._count._all])
      ),
      flaggedConflicts,
      verificationQueueCount,
      bookingsByStatus: Object.fromEntries(
        bookingsByStatus.map((row: { status: string; _count: { _all: number } }) => [row.status, row._count._all])
      ),
      totalUsers,
      totalRevenue: totalRevenue._sum.amount ?? 0,
      refundsPendingDualControl,
      openSupportTicketsByPriority: Object.fromEntries(
        supportTicketsByPriority.map((row: { priority: string; _count: { _all: number } }) => [row.priority, row._count._all])
      ),
    };
  }
}

File: apps/api-py/app/services/dashboards_service.py
"""Direct port of dashboards/dashboards.service.ts."""

from fastapi import HTTPException
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.deps import AuthenticatedUser
from app.models import (
    Booking,
    Favourite,
    OrganisationMember,
    Payment,
    Property,
    Refund,
    SavedSearch,
    SupportTicket,
    Unit,
    User,
)


async def tenant_dashboard(db: AsyncSession, user_id: str) -> dict:
    """Bookings, payments, agreements, support tickets — theirs only."""
    bookings = (
        await db.execute(select(Booking).where(Booking.user_id == user_id).order_by(Booking.created_at.desc()))
    ).scalars().all()
    support_tickets = (
        await db.execute(
            select(SupportTicket).where(SupportTicket.user_id == user_id).order_by(SupportTicket.created_at.desc()).limit(5)
        )
    ).scalars().all()
    favourites_count = (
        await db.execute(select(func.count()).select_from(Favourite).where(Favourite.user_id == user_id))
    ).scalar_one()
    saved_searches_count = (
        await db.execute(select(func.count()).select_from(SavedSearch).where(SavedSearch.user_id == user_id))
    ).scalar_one()

    active_bookings = sum(1 for b in bookings if b.status in ("PENDING_PAYMENT", "CONFIRMED"))

    return {
        "bookings": [{"id": b.id, "status": b.status, "agreedRent": str(b.agreed_rent)} for b in bookings],
        "supportTickets": [{"id": t.id, "subject": t.subject, "status": t.status} for t in support_tickets],
        "counts": {
            "activeBookings": active_bookings,
            "favourites": favourites_count,
            "savedSearches": saved_searches_count,
        },
    }


async def manager_dashboard(db: AsyncSession, user: AuthenticatedUser, organisation_id: str) -> dict:
    """Portfolio composition, verification pipeline status, recent
    bookings, and revenue — scoped to one organisation. Reuses the same
    membership check as properties_service so a manager can never see
    another organisation's numbers."""
    if "Admin" not in user.roles:
        membership = (
            await db.execute(
                select(OrganisationMember).where(
                    OrganisationMember.organisation_id == organisation_id,
                    OrganisationMember.user_id == user.user_id,
                )
            )
        ).scalar_one_or_none()
        if membership is None:
            raise HTTPException(status_code=403, detail="You do not have access to this organisation's dashboard.")

    properties_by_status_rows = (
        await db.execute(
            select(Property.publication_status, func.count())
            .where(Property.organisation_id == organisation_id)
            .group_by(Property.publication_status)
        )
    ).all()

    properties = (await db.execute(select(Property.id).where(Property.organisation_id == organisation_id))).scalars().all()
    total_units = (
        await db.execute(select(func.count()).select_from(Unit).where(Unit.property_id.in_(properties)))
    ).scalar_one() if properties else 0

    recent_bookings = (
        await db.execute(
            select(Booking)
            .join(Unit, Booking.unit_id == Unit.id)
            .join(Property, Unit.property_id == Property.id)
            .where(Property.organisation_id == organisation_id)
            .order_by(Booking.created_at.desc())
            .limit(10)
        )
    ).scalars().all()

    total_revenue = (
        await db.execute(
            select(func.sum(Payment.amount))
            .select_from(Payment)
            .join(Booking, Payment.booking_id == Booking.id)
            .join(Unit, Booking.unit_id == Unit.id)
            .join(Property, Unit.property_id == Property.id)
            .where(Payment.status == "SUCCEEDED", Property.organisation_id == organisation_id)
        )
    ).scalar_one()

    return {
        "propertiesByStatus": {status: count for status, count in properties_by_status_rows},
        "totalProperties": len(properties),
        "totalUnits": total_units,
        "recentBookings": [{"id": b.id, "status": b.status} for b in recent_bookings],
        "totalRevenue": str(total_revenue) if total_revenue else "0",
    }


async def admin_dashboard(db: AsyncSession) -> dict:
    """Platform-wide operational view."""
    properties_by_status_rows = (
        await db.execute(select(Property.publication_status, func.count()).group_by(Property.publication_status))
    ).all()
    flagged_conflicts = (
        await db.execute(select(func.count()).select_from(Property).where(Property.conflict_status == "FLAGGED"))
    ).scalar_one()
    verification_queue_count = (
        await db.execute(select(func.count()).select_from(Property).where(Property.publication_status == "REVIEW"))
    ).scalar_one()
    bookings_by_status_rows = (
        await db.execute(select(Booking.status, func.count()).group_by(Booking.status))
    ).all()
    total_users = (await db.execute(select(func.count()).select_from(User))).scalar_one()
    total_revenue = (
        await db.execute(select(func.sum(Payment.amount)).select_from(Payment).where(Payment.status == "SUCCEEDED"))
    ).scalar_one()
    refunds_pending_dual_control = (
        await db.execute(
            select(func.count()).select_from(Refund).where(Refund.requires_dual_control.is_(True), Refund.approved_by.is_(None))
        )
    ).scalar_one()
    support_by_priority_rows = (
        await db.execute(
            select(SupportTicket.priority, func.count())
            .where(SupportTicket.status.in_(["OPEN", "IN_PROGRESS"]))
            .group_by(SupportTicket.priority)
        )
    ).all()

    return {
        "propertiesByStatus": {status: count for status, count in properties_by_status_rows},
        "flaggedConflicts": flagged_conflicts,
        "verificationQueueCount": verification_queue_count,
        "bookingsByStatus": {status: count for status, count in bookings_by_status_rows},
        "totalUsers": total_users,
        "totalRevenue": str(total_revenue) if total_revenue else "0",
        "refundsPendingDualControl": refunds_pending_dual_control,
        "openSupportTicketsByPriority": {priority: count for priority, count in support_by_priority_rows},
    }
